### PR TITLE
add cgroupv2 oom killer event handling to service management

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: man/systemd.exec.xml [hunk 1]
```diff
@@ -647,11 +647,16 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
...
+        executed processes. Takes an integer between -1000 (to disable OOM killing for this process) and 1000
+        (to make killing of this process under memory pressure very likely). See <ulink
         url="https://www.kernel.org/doc/Documentation/filesystems/proc.txt">proc.txt</ulink> for
-        details.</para></listitem>
+        details.</para>
+
+        <para>Use the <varname>OOMPolicy=</varname> setting of service units to configure how the service
+        manager shall react to the kernel OOM killer terminating a process of the service.  See
+        <citerefentry><refentrytitle>systemd.service</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+        for details.</para></listitem>
```
Comment (boucman) [inline man/systemd.exec.xml:664]: Is it normal that the default value for this setting is not documented ? man:systemd.service tells to look here for the value, but no default value is documented...

Comment (poettering) [reply man/systemd.exec.xml:664]: we should probably document that, yes.

**Comment by poettering** [comment]
Force pushed an updated version. Only change: added the documentation for OOMScoreAdjuts='s default value, as requested by @boucman. (which is strictly speaking unrelated to the rest of this commit, but given that we touch that bit anyway, let's fix this at the same time)

File: src/core/cgroup.c [hunk 2]
```diff
@@ -2479,6 +2555,106 @@ void unit_add_to_cgroup_empty_queue(Unit *u) {
...
+        if (r < 0)
+                return log_unit_debug_errno(u, r, "Failed to parse oom_kill field: %m");
+
+        increased = c > u->oom_kill_last;
+        u->oom_kill_last = c;
+
+        if (!increased)
+                return 0;
+
+        log_struct(LOG_NOTICE,
```
Comment (keszybz) [inline src/core/cgroup.c:2581]: Should this be guarded by a ratelimit check? If a service with lots of tiny processes runs amok, we could be generating quite a lot of messages at notice level.

File: src/core/cgroup.c [hunk 3]
```diff
@@ -2510,15 +2686,16 @@ static int on_cgroup_inotify_event(sd_event_source *s, int fd, uint32_t revents,
                                 /* The watch was just removed */
                                 continue;
 
+                        /* Not that inotify might deliver events for a watch even after it was removed,
```
Comment (keszybz) [inline src/core/cgroup.c]: "note"

File: src/core/cgroup.h [hunk 4]
```diff
@@ -101,6 +101,9 @@ struct CGroupContext {
         uint64_t memory_max;
         uint64_t memory_swap_max;
 
+        /* Configures the memory.oom.group attribute */
+        bool memory_oom_group;
+
```
Comment (keszybz) [inline src/core/cgroup.h]: Maybe stick this with the other `bool`s for pahole friendliness?

File: test/TEST-31-OOMPOLICY/testsuite.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# Let's run this test only if the "memory.oom.group" cgroupfs attribute
+# exists. This test is a bit too strict, since the "memory.events"/"oom_kill"
+# logic has been around since a longer time than "memory.oom.group", but it's
+# an easier thing to test for, and also: let's not get confused by older
+# kernels where the concept was still new.
+
+if test -f /sys/fs/cgroup/system.slice/testsuite.service/memory.oom.group ; then
+
+    systemd-analyze set-log-level debug
+    systemd-analyze set-log-target console
```
Comment (keszybz) [inline test/TEST-31-OOMPOLICY/testsuite.sh]: `systemd-analyze log-level debug`, `systemd-analyze log-target console`

File: test/TEST-31-OOMPOLICY/testsuite.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    while : ; do
+        STATE=`systemctl show -p ActiveState --value oomtest.service`
+        [ "$STATE" = "failed" ] && break
+        sleep .5
+    done
+
+    RESULT=`systemctl show -p Result --value oomtest.service`
+    test "$RESULT" = "oom-kill"
+
+    systemd-analyze set-log-level info
```
Comment (keszybz) [inline test/TEST-31-OOMPOLICY/testsuite.sh]: `systemd-analyze log-level info`

File: src/core/cgroup.c (hunk 2 above)
Comment (poettering) [reply src/core/cgroup.c:2581]: journald does ratelimiting about that anyway. Which is of coursenot that finegrained, but I think good enough for the beginning at least. If people report issues with that we can certainly revisit that though and maybe add something later.

File: src/core/cgroup.h (hunk 4 above)
Comment (poettering) [reply src/core/cgroup.h]: will do

**Comment by poettering** [comment]
Force pushed a new version with the issues pointed out by @keszybz fixed (except the ratelimit thing, see above). Leaving green label on since the changes are so minimal.

**Comment by mrc0mmand** [comment]
Is there any particular reason why the test name is `TEST-31-OOMPOLICY` instead of `TEST-32-OOMPOLICY` (to not clash with an already existing `TEST-31-DEVICE-ENUMERATION`)? Not that it matters, I just found it somewhat odd.

**Pull request merged**